### ISSUE TITLE
compilation error

### ISSUE THREAD

**Issue by drinkingkazu** [LArbys/dllee_unified#3]
Message forwarded from Lauren:

Sorry to bother you all — I know you’re busy. I’m getting another strange (i.e., googling is not very enlightening) error while trying to compile the dllee_unified repository. This one comes from the file Geo2D/Geo2D/PyUtil/Geo2D_PyUtilDict.cxx.

Adrien doesn’t get the same error on his machine, so there’s a reasonably high probability that it’s something messed up within my own setup. Any thoughts would be appreciated.

Compiling PyUtil...
<< compiling Geo2D_PyUtilDict.cxx >>
In file included from Geo2D_PyUtilDict.cxx:40:
In file included from ./PyDraw.h:33:
In file included from /usr/local/include/opencv2/opencv.hpp:52:
In file included from /usr/local/include/opencv2/core.hpp:3281:
In file included from /usr/local/include/opencv2/core/cvstd.inl.hpp:47:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/complex:247:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:138:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:478:15: error: 
     C++ requires a type specifier for all declarations
   char_type toupper(char_type __c) const
             ^
/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/pyport.h:731:29: note: 
     expanded from macro 'toupper'
#define toupper(c) towupper(btowc(c))
                           ^
In file included from Geo2D_PyUtilDict.cxx:40:
In file included from ./PyDraw.h:33:
In file included from /usr/local/include/opencv2/opencv.hpp:52:
In file included from /usr/local/include/opencv2/core.hpp:3281:
In file included from /usr/local/include/opencv2/core/cvstd.inl.hpp:47:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/complex:247:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:138:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:484:48: error: 
     too many arguments provided to function-like macro invocation
   const char_type* toupper(char_type* __low, const char_type* __high) const
                                              ^
/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/pyport.h:731:9: note: 
     macro 'toupper' defined here
#define toupper(c) towupper(btowc(c))
       ^
In file included from Geo2D_PyUtilDict.cxx:40:
In file included from ./PyDraw.h:33:
In file included from /usr/local/include/opencv2/opencv.hpp:52:
In file included from /usr/local/include/opencv2/core.hpp:3281:
In file included from /usr/local/include/opencv2/core/cvstd.inl.hpp:47:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/complex:247:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:138:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:484:29: error: 
     expected ';' at end of declaration list
   const char_type* toupper(char_type* __low, const char_type* __high) const
                           ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:496:48: error: 
     too many arguments provided to function-like macro invocation
   const char_type* tolower(char_type* __low, const char_type* __high) const
                                              ^
/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/pyport.h:729:9: note: 
     macro 'tolower' defined here
#define tolower(c) towlower(btowc(c))
       ^
In file included from Geo2D_PyUtilDict.cxx:40:
In file included from ./PyDraw.h:33:
In file included from /usr/local/include/opencv2/opencv.hpp:52:
In file included from /usr/local/include/opencv2/core.hpp:3281:
In file included from /usr/local/include/opencv2/core/cvstd.inl.hpp:47:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/complex:247:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:138:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:480:27: error: 
     use of undeclared identifier '__c'
       return do_toupper(__c);
                         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:587:15: error: 
     C++ requires a type specifier for all declarations
   char_type toupper(char_type __c) const
             ^
/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/pyport.h:731:29: note: 
     expanded from macro 'toupper'
#define toupper(c) towupper(btowc(c))
                           ^
In file included from Geo2D_PyUtilDict.cxx:40:
In file included from ./PyDraw.h:33:
In file included from /usr/local/include/opencv2/opencv.hpp:52:
In file included from /usr/local/include/opencv2/core.hpp:3281:
In file included from /usr/local/include/opencv2/core/cvstd.inl.hpp:47:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/complex:247:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:138:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:593:48: error: 
     too many arguments provided to function-like macro invocation
   const char_type* toupper(char_type* __low, const char_type* __high) const
                                              ^
/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/pyport.h:731:9: note: 
     macro 'toupper' defined here
#define toupper(c) towupper(btowc(c))
       ^
In file included from Geo2D_PyUtilDict.cxx:40:
In file included from ./PyDraw.h:33:
In file included from /usr/local/include/opencv2/opencv.hpp:52:
In file included from /usr/local/include/opencv2/core.hpp:3281:
In file included from /usr/local/include/opencv2/core/cvstd.inl.hpp:47:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/complex:247:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:138:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:593:29: error: 
     expected ';' at end of declaration list
   const char_type* toupper(char_type* __low, const char_type* __high) const
                           ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:605:48: error: 
     too many arguments provided to function-like macro invocation
   const char_type* tolower(char_type* __low, const char_type* __high) const
                                              ^
/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/pyport.h:729:9: note: 
     macro 'tolower' defined here
#define tolower(c) towlower(btowc(c))
       ^
In file included from Geo2D_PyUtilDict.cxx:40:
In file included from ./PyDraw.h:33:
In file included from /usr/local/include/opencv2/opencv.hpp:52:
In file included from /usr/local/include/opencv2/core.hpp:3281:
In file included from /usr/local/include/opencv2/core/cvstd.inl.hpp:47:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/complex:247:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:138:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:589:27: error: 
     use of undeclared identifier '__c'
       return do_toupper(__c);
                         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:715:21: error: 
     too many arguments provided to function-like macro invocation
isspace(_CharT __c, const locale& __loc)
                   ^
/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/pyport.h:725:9: note: 
     macro 'isspace' defined here
#define isspace(c) iswspace(btowc(c))
       ^
In file included from Geo2D_PyUtilDict.cxx:40:
In file included from ./PyDraw.h:33:
In file included from /usr/local/include/opencv2/opencv.hpp:52:
In file included from /usr/local/include/opencv2/core.hpp:3281:
In file included from /usr/local/include/opencv2/core/cvstd.inl.hpp:47:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/complex:247:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:138:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:715:1: error: 
     declaration conflicts with target of using declaration already in scope
isspace(_CharT __c, const locale& __loc)
^
/usr/include/ctype.h:267:1: note: target of using declaration
isspace(int _c)
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cctype:113:9: note: 
     using declaration
using ::isspace;
       ^
In file included from Geo2D_PyUtilDict.cxx:40:
In file included from ./PyDraw.h:33:
In file included from /usr/local/include/opencv2/opencv.hpp:52:
In file included from /usr/local/include/opencv2/core.hpp:3281:
In file included from /usr/local/include/opencv2/core/cvstd.inl.hpp:47:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/complex:247:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:138:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:717:5: error: 
     expected expression
   return use_facet<ctype<_CharT> >(__loc).is(ctype_base::space, __c);
   ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:718:2: error: 
     expected ';' at end of declaration
}
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:739:21: error: 
     too many arguments provided to function-like macro invocation
isupper(_CharT __c, const locale& __loc)
                   ^
/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/pyport.h:727:9: note: 
     macro 'isupper' defined here
#define isupper(c) iswupper(btowc(c))
       ^
In file included from Geo2D_PyUtilDict.cxx:40:
In file included from ./PyDraw.h:33:
In file included from /usr/local/include/opencv2/opencv.hpp:52:
In file included from /usr/local/include/opencv2/core.hpp:3281:
In file included from /usr/local/include/opencv2/core/cvstd.inl.hpp:47:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/complex:247:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:138:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:739:1: error: 
     declaration conflicts with target of using declaration already in scope
isupper(_CharT __c, const locale& __loc)
^
/usr/include/ctype.h:273:1: note: target of using declaration
isupper(int _c)
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cctype:114:9: note: 
     using declaration
using ::isupper;
       ^
In file included from Geo2D_PyUtilDict.cxx:40:
In file included from ./PyDraw.h:33:
In file included from /usr/local/include/opencv2/opencv.hpp:52:
In file included from /usr/local/include/opencv2/core.hpp:3281:
In file included from /usr/local/include/opencv2/core/cvstd.inl.hpp:47:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/complex:247:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:138:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:741:5: error: 
     expected expression
   return use_facet<ctype<_CharT> >(__loc).is(ctype_base::upper, __c);
   ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:742:2: error: 
     expected ';' at end of declaration
}
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:747:21: error: 
     too many arguments provided to function-like macro invocation
islower(_CharT __c, const locale& __loc)
                   ^
/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/pyport.h:723:9: note: 
     macro 'islower' defined here
#define islower(c) iswlower(btowc(c))
       ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[1]: *** [Geo2D_PyUtilDict.o] Error 1
make: *** [all] Error 2

**Comment by drinkingkazu** [comment]
@lyates17  I'm probably most out-dated = least helpful. How did you install OpenCV? It looks like there's a possible mismatch between your opencv and compiler?

**Comment by lyates17** [comment]
@drinkingkazu I installed OpenCV using homebrew -- following the instructions in the dllee_unified README.md. However, looking at the details of the install now, I see it's opencv 3.4.0_1 whereas the desired version for dllee_unified is 3.0.

**Comment by drinkingkazu** [comment]
@lyates17 assuming:

0. you already tried "make clean; rm -rf build; make" and that did not solve
1. others don't give an easier suggestion to try than the bullet below
2. try installing opencv 3.2 or 3.0 (I used to use 3.2).

**Comment by lyates17** [comment]
0. I did "make clean" and "make" within the Geo2D directory. There is no "build" directory that I can see within Geo2D or any of its subdirectories, so I didn't do "rm -rf build" in between. The error persists after taking these steps.
1. No other suggestions yet...
2. I will start looking into installing a previous version of opencv unless I hear other ideas. I will report back on whether I am able to do that and whether it resolves this issue.

**Comment by lyates17** [comment]
In the process of looking into installing a previous version of opencv, I learned that homebrew has fairly recently changed their handling of homebrew vs. the Apple system python install. The should version is that I needed to add an extra line to my ~/.bash_profile to get brew's python executable into my PATH (in addition the usual `export PATH="/usr/local/bin:$PATH"`). A brief discussion related to this to this issue can be found in the "Caveats" section of the output from `brew info python`.

I added the specified line to my ~/.bash_profile, opened a new terminal, and sourced clean.sh (just to be safe). After that, I was able to source build.sh and build successfully (although still sans the PMT precuts and LArLiteCV). Closing this.
